sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java

```diff
@@ -160,40 +175,96 @@ public FunctionImplementation getForTypes(List<DataType> args) throws IllegalArg
+                        break;
 
-                case LongType.ID:
-                case TimestampType.ID_WITH_TZ:
-                case TimestampType.ID_WITHOUT_TZ:
-                    scalar = new BinaryScalar<>(longFunction, name, DataTypes.LONG, features);
-                    break;
+                    case FloatType.ID:
+                        scalar = new BinaryScalar<>(floatFunction, name, DataTypes.FLOAT, features);
+                        break;
 
-                default:
-                    throw new UnsupportedOperationException(
-                        operator + " is not supported on expressions of type " + dataType.getName());
+                    case ByteType.ID:
+                    case ShortType.ID:
+                    case IntegerType.ID:
+                        scalar = new BinaryScalar<>(integerFunction, name, DataTypes.INTEGER, features);
+                        break;
+
+                    case IntervalType.ID:
+                        scalar = new IntervalArithmeticScalar(operator);
+                        break;
+
+                    case LongType.ID:
+                    case TimestampType.ID_WITH_TZ:
+                    case TimestampType.ID_WITHOUT_TZ:
+                        scalar = new BinaryScalar<>(longFunction, name, DataTypes.LONG, features);
+                        break;
+
+                    default:
+                        throw new UnsupportedOperationException(
+                            operator + " is not supported on expressions of type " + fst.getName());
+                }
+                return Scalar.withOperator(scalar, operator);
+            }
+
+            if (isInterval(fst) && isTimestamp(snd)) {
+                return new IntervalTimestampScalar(operator, fst, snd, snd);
+            }
+            if (isTimestamp(fst) && isInterval(snd)) {
+                return new IntervalTimestampScalar(operator, fst, snd, fst);
+            }
+            if (isInterval(fst) && isNumerical(snd)) {
+                return new IntervalNumberScalar(operator, snd);
             }
-            return Scalar.withOperator(scalar, operator);
+
+            throw new UnsupportedOperationException(
+                String.format(Locale.ENGLISH, "Arithmetic operation are not supported for type %s %s", fst, snd));
+        }
+
+        private static boolean isNumerical(DataType d) {
+            return NUMERICAL_ID.contains(d.id());
+        }
+
+        private static boolean isInterval(DataType d) {
+            return d.id() == IntervalType.ID;
         }
+
+        private static boolean isTimestamp(DataType d) {
+            return TIMESTAMP_IDS.contains(d.id());
+        }
+
+        private static final Set<Integer> TIMESTAMP_IDS = Sets.newHashSet(DataTypes.TIMESTAMP.id(),
+                                                                          DataTypes.TIMESTAMPZ.id());
+
+        private static final Set<Integer> NUMERICAL_ID = Sets.newHashSet(DataTypes.DOUBLE.id(),
```

Comment:
I think we've already a definition like this in  `DataTypes` which also includes short, can we use that instead?